sympy/tensor/array/array_comprehension.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import print_function, division
+import functools
+from sympy.core.sympify import sympify
+from sympy.core.expr import Expr
+from sympy.core import Basic
+from sympy.core.compatibility import Iterable
+from sympy.tensor.array import MutableDenseNDimArray, ImmutableDenseNDimArray
+from sympy import  Symbol
+from sympy.core.sympify import sympify
+from sympy.core.compatibility import Iterable
+from sympy.core.numbers import Integer
+
+class ArrayComprehension(Basic):
+    """
+    Generate a list comprehension
+    If there is a symbolic dimension, for example, say [i for i in range(1, N)] where
+    N is a Symbol, then the expression will not be expanded to an array. Otherwise,
+    calling the doit() function will launch the expansion.
+
+    Examples
+    ========
+
+    >>> from sympy.tensor.array import ArrayComprehension
+    >>> from sympy.abc import i, j, k
+    >>> a = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, 3))
+    >>> a
+    ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, 3))
+    >>> a.doit()
+    [[11, 12, 13], [21, 22, 23], [31, 32, 33], [41, 42, 43]]
+    >>> b = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, k))
+    >>> b.doit()
+    ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, k))
+
+    """
+    def __new__(cls, expr, *bounds, **assumptions):
+        if any(len(l) != 3 or None for l in bounds):
+            raise ValueError('ArrayComprehension requires values lower and upper bound'
+                              ' for the expression')
+        cls.default_assumptions = assumptions
+        arglist = [sympify(expr)]
+        arglist.extend(cls._check_bounds_validity(expr, bounds))
+        obj = Basic.__new__(cls, *arglist, **assumptions)
+        obj._expr = arglist[0]
+        obj._bounds = arglist[1:]
+        return obj
+
+    @property
+    def expr(self):
+        """
+        Return the expression that will be expanded to an array
+
+        Examples
+        ========
+
+        >>> from sympy.tensor.array import ArrayComprehension
+        >>> from sympy.abc import i, j, k
+        >>> a = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, 3))
+        >>> a.expr
+        10*i + j
+        """
+        return self._args[0]
+
+    @property
+    def bounds(self):
```

Comment:
is this the same terminology used in `Sum` and `Product`? I remember a different name, can you check?

Reply:
In `Sum` and `product`,  `function` ans `symbols` are used as parameters of construtor which correspond to `expr` and `bounds`. But then `symbols` is transformed `limits`. Since there is not a standard naming convention, I created the name by my own.
Besides, there are not explicit properties in `Sum` to make these parameters accesible. In `product`， there is only one for `function`, but it is renamed as `term`. 
In this case, what terminology do you think I should keep?

Reply:
I think there is a standard terminology. There is a superclass implementing it.

Reply:
@Upabjojr 
I think what you are refering to is the `ExprWithLimits` class. And yes, it defines many properties like `function`, `limits`, `variables`, `bound_symbols`, `free_symbols`, `is_number`. Do you prefer folowing these examples?

Reply:
Maybe yes, if no complications arise I'm +1 with that. You can do this here or in another PR.

Reply:
Maybe just make sure the names correspond with `ExprWithLimits`, I want this PR to be merged as soon as possible. Subclassing can be done in the future.